queries.js

```diff
@@ -145,291 +145,250 @@ function parseShesmuType(type) {
   }
 }
 
+/** set up custom error if bad params are given */
+function ValidationError(message) {
+  this.name = 'ValidationError';
+  this.message = message || '';
+}
+ValidationError.prototype = Error.prototype;
+
 /**
  * validate parameters against permitted parameters
  */
-function validateParameters(req, res, permitted_parameters) {
-  var permittedNames = Object.keys(permitted_parameters);
-  for (var nameIndex = 0; nameIndex < permittedNames.length; nameIndex++) {
-    var name = permittedNames[nameIndex];
+function throwIfHasInvalidParams(req, res, permitted_parameters) {
+  const permittedNames = Object.keys(permitted_parameters);
+  for (let nameIndex = 0; nameIndex < permittedNames.length; nameIndex++) {
+    const name = permittedNames[nameIndex];
     if (req.body.parameters.hasOwnProperty(name)) {
       let type = parseShesmuType(permitted_parameters[name].type);
       if (!type.check(req.body.parameters[name])) {
-        res
-          .status(400)
-          .send(
-            'Invalid value for parameter ' +
-              name +
-              '. Expected type ' +
-              permitted_parameters[name].type +
-              '.'
-          );
-        return 0;
+        throw new ValidationError(
+          'Invalid value for parameter ' +
+            name +
+            '. Expected type ' +
+            permitted_parameters[name].type +
+            '.'
+        );
       }
       req.body.parameters[name] = type.canonicalise(req.body.parameters[name]);
     } else if (permitted_parameters[name].required) {
-      res
-        .status(400)
-        .send('Report requires ' + name + ', but paramter not supplied.');
-      return 0;
+      throw new ValidationError(
+        'Report requires ' + name + ', but it was not supplied.'
+      );
     }
   }
-  var providedNames = Object.keys(req.body.parameters);
-  for (nameIndex = 0; nameIndex < providedNames.length; nameIndex++) {
-    name = providedNames[nameIndex];
+  const providedNames = Object.keys(req.body.parameters);
+  for (let nameIndex = 0; nameIndex < providedNames.length; nameIndex++) {
+    const name = providedNames[nameIndex];
     if (!permitted_parameters.hasOwnProperty(name)) {
-      res.status(400).send('Invalid parameter ' + name + ' provided.');
-      return 0;
+      throw new ValidationError('Invalid parameter ' + name + ' provided.');
     }
   }
   return 1;
 }
 
 // add query functions
 
-function getAllReports(req, res, next) {
-  db
-    .any('select * from report')
-    .then(function(data) {
-      res.status(200).json(data);
-    })
-    .catch(function(err) {
-      return next(err);
-    });
+async function getAllReports(req, res, next) {
```

Comment:
This is a convention specific to the Express framework. As part of its routing, it calls the function with the arguments `(req, res, next)` in that order, so that's why it has to be passed as an argument even though it isn't necessarily used.